### PR TITLE
Modules

### PR DESCRIPTION
**Pull request by karthiknadig**
Using this PR to discus the right format and any performance implications of searching for modules from stack traces.

In this PR:
1. Handles modules request #21
2. Generate modules events based on modules seen in stack frames #22
3. *Does NOT report unloaded modules
4. Formats stack frames based on modules #64 (partial)

This is with **modules** and **lines** on:
image

This is with only **lines** on:
image

This is with nothing on:
image

Pending:

- [x] Update to stack trace tests to include formatting
- [x] New tests for modules request
- [x] Fixes for broken tests

### CODE REVIEW

**Comment by zooba** [comment]
Can we format with full dotted names like you'd import them? I don’t think the length problem will be worse than in any other language, and we ought to be consistent as much as we can.

File: ptvsd/wrapper.py [hunk 1]
```diff
@@ -508,6 +508,68 @@ def get_sort_key(o):
...
+            if path and mpath == path:
+                with self._lock:
+                    module_id = self._next_id
+                    self._next_id += 1
+
+                    self.path_to_module_id[module_path] = module_id
+
+                    module = {
+                        'id': module_id,
+                        'name': value.__name__,
```
Comment (int19h) [inline ptvsd/wrapper.py]: Let's use `__qualname__` here if it's available, and fall back to `__name__` otherwise.

File: ptvsd/wrapper.py [hunk 2]
```diff
@@ -508,6 +508,68 @@ def get_sort_key(o):
...
+
+                    self.module_id_to_details[module_id] = module
+
+                    self.proc.send_event('module', reason='new', module=module)
+                    return module
+
+        return None
+
+    def get_all(self):
+        for _, value in self.module_id_to_details.items():
```
Comment (int19h) [inline ptvsd/wrapper.py]: You can use `values()` instead of `items()` to iterate just through those.

File: ptvsd/wrapper.py (hunk 1 above)
Comment (int19h) [inline ptvsd/wrapper.py]: `__qualname__` is also interesting when it's available (and can then be used in `__format_frame_name`).

**Comment by karthiknadig** [comment]
@zooba, @int19h is this better:
```
__main__ : 60
_pydev_imps._pydev_execfile.execfile : 25
pydevd.run : 1035
pydevd.main : 1628
ptvsd.debugger._run : 54
ptvsd.debugger._run_file : 36
ptvsd.debugger.debug : 24
pydevd : 96  <---- this one is wrong but that is because python thinks ptvsd_launcher.py is pydevd module
```

We currently don't have a good way to get fully qualified function names. We could use custom messages to talk to analyzer for more detailed function info (including parameter type, name, and values).

**Comment by int19h** [comment]
For function names, can we reuse `QualifiedVersionNameWalker` on VS side to improve the output? We could do it before, because we were providing `AD7StackFrame` ourselves. Not sure if it's possible with the adapter.

(Not for this PR, just an idea.)

**Comment by int19h** [comment]
@ericsnowcurrently, can you take a look at the tests?

**Comment by ericsnowcurrently** [comment]
Yeah, but it will probably be Monday.  I'm still at the DevDiv bootcamp.

**Comment by int19h** [comment]
I'd like to get this in today so that it can go into the VS bundle - this is one of those things that users will notice right away. Let's go ahead with the merge, but feel free to leave any comments on how the tests can be improved, and Karthik can take care of that next week.

**Comment by ericsnowcurrently** [comment]
will do

**Pull request merged**